rust/main/agents/relayer/src/msg/metadata/message_builder.rs

```diff
@@ -0,0 +1,430 @@
+        base_builder
+    }
+
+    fn insert_mock_routing_ism(base_builder: &MockBaseMetadataBuilder) {
+        let mock_ism = MockInterchainSecurityModule::default();
+        mock_ism
+            .responses
+            .module_type
+            .lock()
+            .unwrap()
+            .push_back(Ok(ModuleType::Routing));
+        base_builder
+            .responses
+            .build_ism
+            .lock()
+            .unwrap()
+            .push_back(Ok(Box::new(mock_ism)));
+
+        let routing_ism = MockRoutingIsm::default();
+        routing_ism
+            .responses
+            .route
+            .lock()
+            .unwrap()
+            .push_back(Ok(H256::zero()));
+        base_builder
+            .responses
+            .build_routing_ism
+            .lock()
+            .unwrap()
+            .push_back(Ok(Box::new(routing_ism)));
+    }
+
+    #[tokio::test]
+    async fn depth_already_reached() {
+        let base_builder = build_mock_base_builder();
+        insert_mock_routing_ism(&base_builder);
+
+        let ism_address = H256::zero();
+        let message = HyperlaneMessage::default();
+
+        let mut message_builder =
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+
+        message_builder.depth = ISM_MAX_DEPTH;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmDepthExceeded(ISM_MAX_DEPTH));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn ism_count_already_reached() {
+        let base_builder = build_mock_base_builder();
+        insert_mock_routing_ism(&base_builder);
+
+        let ism_address = H256::zero();
+        let message = HyperlaneMessage::default();
+
+        let message_builder =
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+
+        *message_builder.ism_count.lock().await = ISM_MAX_COUNT;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmCountReached(ISM_MAX_COUNT));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn max_depth_reached() {
```

Comment:
i'd add a check for both the max_depth and max_count in all unit tests so the reader can more easily understand the ISM layout. Or maybe the depth and count can be passed as parameters to the initialization function